### PR TITLE
`General`: Add feature toggle for science api

### PR DESCRIPTION
**Pull request by MaximilianAnzinger**
> [!WARNING]
> **This PR is stacked on #7680 ! Do not merge this PR!**
> **Do only test on TS1!**

#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The base PR for the science API (#7680) doesn't include a feature toggle to allow administrators to disable the logging of events.

### Description

This PR adds the new feature toggle for science API.

### Steps for Testing

Prerequisites:
- 1 Admin
- 1 Course with a lecture

1. Log in to Artemis
2. Navigate to Server Administration > Feature Toggles
3. Make sure the Science Feature is disabled per default
4. Navigate to User Settings > Science
5. Make sure you can see the message, that the feature is currently disabled
6. Navigate to Course > Lectures and view the lecture details page
7. Check in the database that no event was logged for the user
8. Navigate to Server Administration > Feature Toggles
9. Enable the Science Feature
10. Navigate to User Settings > Science
11. Make sure you don't see the previous message
12. Navigate to Course > Lectures and view the lecture details page
13. Check in the database that the event was logged for the user

Config Test:
1. Start the Server with a config that has `artemis.science.event-logging.enable: false`
2. Log in as Admin
3. Navigate to feature toggles and make sure the science toggle is disabled
4. Start the Server with a config that has `artemis.science.event-logging.enable: true`
5. Log in as Admin
6. Navigate to feature toggles and make sure the science toggle is enabled

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [feature-toggle.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6042/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/feature-toggle/feature-toggle.service.ts.html) | 66.66% | ✅  |
| [science.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6042/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/science/science.service.ts.html) | 92.59% | ✅  |
| [science-settings.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6042/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/user-settings/science-settings/science-settings.component.ts.html) | 90.9% | ✅  |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [Feature.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6042/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.feature/Feature.html) | 100% | ✅ |
| [FeatureToggleService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6042/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.feature/FeatureToggleService.html) | 100% | ✅ |
| [ScienceResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6042/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.science/ScienceResource.html) | 100% | ✅ |

### Screenshots

Screenshot 2024-01-27 at 20 44 36

### CODE REVIEW

**Review by Predixx — COMMENTED** [review]
I tested the science feature and all the text is working. I just couldn't check the db (no access)

**Comment by MaximilianAnzinger** [comment]
> As far as I understand the feature toggle would have to be flipped manually by an admin after each server restart since it is not persistent. Would it make sense to introduce a configuration option where admins can define if it should be enabled or disabled by default?

I like your suggestion. I added it and set the default to false 👍

File: src/main/resources/config/application-artemis.yml [hunk 1]
```diff
@@ -109,3 +109,6 @@ artemis:
 
     plagiarism-checks:
         plagiarism-results-limit: 100
+
+    science:
+        enable-on-start: false # Set this to true to enable the feature toggle for the science module on startup.
```
Comment (b-fein) [inline src/main/resources/config/application-artemis.yml:114]:
```suggestion
    research:
        event-logging:
            enable: false
```
I think that an admin that did not read the PR here would have no idea what ‘enable science’ means.

The proposed config key would also allow for potential extension in the future, e.g.,
```yaml
research:
  # enable base functionality globally, then allow more fine-grained selection of events that are collected
  event-logging:
    enable: true
    lecure-events: true
    programming-exercise-events: false
  # some other research-functionality options here in the future?
```

File: src/main/java/de/tum/in/www1/artemis/service/feature/FeatureToggleService.java [hunk 2]
```diff
@@ -14,6 +17,9 @@ public class FeatureToggleService {
 
     private static final String TOPIC_FEATURE_TOGGLES = "/topic/management/feature-toggles";
 
+    @Value("${artemis.science.enable-on-start:false}")
+    private Boolean scienceEnabledOnStart;
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/feature/FeatureToggleService.java:21]:
```suggestion
    private boolean scienceEnabledOnStart;
```

**Review by rstief — DISMISSED** [review]
works as expected

**Review by reschandreas — APPROVED** [review]
feature toggle can be disabled as expected

**Review by basak-akan — DISMISSED** [review]
Tested during the testing session, worked well

**Review by rstief — APPROVED** [review]
tested locally, works as expected.

**Comment by MaximilianAnzinger** [comment]
> How can one check in the database that the event was logged for the user?

```sql
SELECT * FROM artemis.science_event
```

**Review by bassner — APPROVED** [review]
Works fine on TS1. LGTM!

**Pull request merged**